### PR TITLE
Polish the checkbox, update switch documentation

### PR DESCRIPTION
**Pull request by jasmussen**
The switch has a long history. Before I go into details, here's what this PR does:

- It replaces "Pending Review", "Stick to Front Page", "Allow Comments", and "Allow Pingbacks" from using Switches to using Checkboxes.
- It tweaks the design of the checkbox to have a significantly more visible checked state, and a new more visual focus state.
- It tweaks the documentation for when to use a switch instead of a checkbox.
- It unifies the margins of checkboxes to use our 4/8/12 grid, and adds a little padding.

The Switch is born from mobile, which is increasingly the device of choice for people and long term might become the only choice. It's bigger and more tappable, and implies by its nature that it's a boolean choice. It also has qualities that the checkbox does not embody:

- Like its wall lightswitch counterpart, it’s something you flip and the effect is instant. There used to be darkness, now there’s light.
- It has only a single purpose, and that purpose is boolean. For example you wouldn’t use switches to select multiple categories.

However "Allow Comments" and the three other aforementioned toggles that use switches in master, do not have an instant effect, they require you to hit "Save" before the toggle option is stored. Given the above reasoning, a checkbox is a better choice in those situations.

The toggle is still a good choice in situations where immediate visual feedback is given, like when a dropcap is applied, or when you toggle image croppping for gallery items. In those situations, the pairing of the boolean nature and the larger touch target makes the switch feel more substantial.

With this PR I'm trying to take into account all the past discussions on the topic, and I know there have been many. Lay your thoughts on me.

Screenshots, checking and focus styles:

styles

Relaxed margins:

tweaks

Instant effect toggles:

switch

### CODE REVIEW

File: edit-post/assets/stylesheets/main.scss
Comment (youknowriad) [inline edit-post/assets/stylesheets/main.scss]: Should we move these styles to the "Checkbox control" or do we want them to apply globally to checkboxes?

 - Moving to CheckboxControl means, it works outside Gutenberg, outside WordPress, whenever you use the component
 - It also means that if you use a checkbox with the `CheckboxControl` component, it won't have the styles. 

So each option has pros/cons.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [inline edit-post/assets/stylesheets/main.scss]: "ones" is a *slightly* ambiguous word here; it'd say "styles" would be better.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [inline edit-post/assets/stylesheets/main.scss]: Can you file an issue for this please and link to it? These TODOs will get lost to time otherwise 😄

At least if there's an issue it'll get lost to time but be tracked! 😉

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [inline edit-post/assets/stylesheets/main.scss]: That hardcoded border width makes me nervous. Do we want a rule that checkboxes have double the usual border width? I dunno.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [inline edit-post/assets/stylesheets/main.scss]: This would be more readable if it were underneath the plain `:checked` style below. I think it's nice to "stack" styles like that.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [inline edit-post/assets/stylesheets/main.scss]: Isn't this selector technically supposed to be `::before`? I know `:before` works but we should use `::before`: https://developer.mozilla.org/en-US/docs/Web/CSS/::before#Syntax

We should probs have a rule for it 🤷‍♂️

File: packages/components/src/form-toggle/README.md
Comment (tofumatt) [inline packages/components/src/form-toggle/README.md]: I have a few tweaks to make here but I'll just push a commit 😄

**Review by tofumatt — COMMENTED** [review]
I largely dig it, I think a few tweaks would be nice but I'll push a few things and let you respond to feedback... after that it's good to go.

**Comment by tofumatt** [comment]
Also, thanks for the fantastic PR summary, really made this one a breeze to review 😄

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [reply edit-post/assets/stylesheets/main.scss]: Shoot, I thought I commented on this but didn't hit submit. Was gonna say the same thing, I don't think this should be a global checkbox style. We should scope it to our component and just make sure we're using the `CheckboxControl`.

**Comment by paulwilde** [comment]
By changing them to using the `CheckboxControl` the `withInstanceId` can now be dropped to clean up a bunch of these.

File: edit-post/assets/stylesheets/main.scss
Comment (jasmussen) [reply edit-post/assets/stylesheets/main.scss]: So far I've intentionally not filed, because the styles for these components were partially in flux. But as of the most recent input style change, and now this, things are solidifiying. How about this: I await a little feedback on this branch, and if the thumbs point upwards, I will create the ticket, update the PR with a link in the comment, and we'll go from there?

File: edit-post/assets/stylesheets/main.scss
Comment (jasmussen) [reply edit-post/assets/stylesheets/main.scss]: Hardcoded makes it sound so serious :) 

The double border is to give weight to the checkbox, without making it overwhelming. Visually I like the balance but design feedback may change this. What would be good CSS if/when the visuals are settled?

**Comment by tofumatt** [comment]
(Really sorry but I got a bit distracted by another branch so I'll get back to this in a few hours, hope that's okay.)

**Comment by jasmussen** [comment]
No worries, I'm about to push a few fixes based on your feedback so maybe hold off a minute, just npm installing a thing.

File: edit-post/assets/stylesheets/main.scss
Comment (jasmussen) [reply edit-post/assets/stylesheets/main.scss]: Pushing a fix in a second, let me know if that actually addresses your comment, wasn't sure I understood you perfectly.

File: edit-post/assets/stylesheets/main.scss
Comment (jasmussen) [reply edit-post/assets/stylesheets/main.scss]: It is, and I know, and I'm pretty sure we've discussed this. The reason I went with single colon (besides habits, but habits can change, you've seen this from me capitalizing and period ending my comments now), is that single colon is what we've used EVERYWHERE else in the code. If we add a linter rule and convert those all to double colons, that could probably be a good separate PR no?

**Comment by jasmussen** [comment]
Okay, pushed a fix for some of the comments, be sure to pull if you wanted to change the documentation!

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [reply edit-post/assets/stylesheets/main.scss]: That's legit. Filed #8618.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [reply edit-post/assets/stylesheets/main.scss]: Yeah that makes sense. I just think either a variable named 2px or even just using `$border-width + 1` would work? That's what this is... and it makes sense to base it off our global borders.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [reply edit-post/assets/stylesheets/main.scss]: Works for me. In general I say more issues (especially things like this which are easily marked as "code quality" issues we can filter out). Easier to close unneeded ones than remember long-forgotten ones 😉

**Comment by tofumatt** [comment]
Thanks, sorry I got distracted, my brain is scattered all over the map today.

**Review by karmatosed — APPROVED** [review]
I really like this! Thanks for tackling this issue.

File: edit-post/assets/stylesheets/main.scss
Comment (tofumatt) [reply edit-post/assets/stylesheets/main.scss]: > If there's a JavaScripty future for the rest of WordPress

I think there is 😄

-----

I think we should try to contain our styles and apply them only to our components. We should simply get more things using our components, but the benefits of component-based development is that we can scope styles to an easily-reusable component and then just use it everywhere.

I'd really prefer we scope the styles to this component.

**Review by tofumatt — CHANGES_REQUESTED** [review]
I pushed my docs tweak, but I think we should be scoping our styles and I wonder about that TODO now that design feedback is addressed.

So I think a few things here need to be done before we merge, but largely I dig it. Ping me if I was unclear about anything.

**Comment by afercia** [comment]
For reference, there's a pending ticket for core (3 years) proposing to improve the checkboxes (and radio buttons) contrast ratio. This new Gutenberg design seems to be a very good improvement. https://core.trac.wordpress.org/ticket/35596

File: edit-post/assets/stylesheets/main.scss
Comment (jasmussen) [reply edit-post/assets/stylesheets/main.scss]: @tofumatt I appreciate your passion for code cleanliness. I share it. My favorite t-shirt is one my friend @MichaelArestad wore a few years ago, that spells "Hella Clean Code" with sparkles around it. 

In this case, however, if I scope the checkbox component to be _just scoped to the component_, the categories aren't styled:

screen shot 2018-08-07 at 11 22 52

I'm guessing that can be fixed by actually using the checkbox control as part of the categories component, and that this in turn would be the right thing to do. But you'll notice that <a href="https://github.com/WordPress/gutenberg/blob/a8f5319995e4aa8c281390e75cb2dd365c2e087d/edit-post/assets/stylesheets/main.scss#L130">the input styles</a> starting on this line are also located in this very same file. Should they also be moved to components? And in turn, which input styles will then break if we do this?

I think an argument can be made that the these form widgets, `input`, `select`, checkboxes, etc. are sort of base ingredients that should be styled in a global stylesheet, as opposed to on a scoped component basis. Or, should each of those HTML elements be a component on its own? And if so, is that a separate PR?

I'm not trying to be difficult — truly my biggest goal here is to simply make sure that all the textfields and checkboxes inside Gutenberg look consistent, and longer term that those same styles can be removed from Gutenberg because they are now default in WordPress as a whole (hence the need for the upstream tickets).

What's the best path forward?

Comment (tofumatt) [reply edit-post/assets/stylesheets/main.scss]: Hmmm... okay. So I'd say that what we really want is:

* styles are *always scoped to a component
* we use a component for everything (even making our own `checkbox` component so it has styles)

That said, you're right for now: it's too much of a pain to do that and it's out of scope.

Your PR is an improvement. Let's not let me make perfect the enemy of great! 😄

**Review by tofumatt — APPROVED** [review]

**Comment by jasmussen** [comment]
I created https://core.trac.wordpress.org/ticket/40633 to track the new styles, and replaced the todos in https://github.com/WordPress/gutenberg/pull/8588/commits/9d3aa1859ba193f9d4f619762a409b4751725b39. 

Thanks for the reviews everyone, I will merge if the tests pass.

**Comment by jasmussen** [comment]
The tests pass for me locally, but I'm not sure why travis keeps failing. I've tried restarting it twice now. Any insights?

**Comment by jasmussen** [comment]
I'm getting this e2e test:

screen shot 2018-08-08 at 14 00 43

But I'm not sure why that should be affected by this branch, as it's mostly touching CSS and a few panels in the inspector. Is this failing in master too?

**Comment by gziolo** [comment]
`test/e2e/specs/writing-flow.test.js` fails also locally for me on the latest master branch. Let's merge it as is 👍

**Pull request merged**
